Add ML-KEM/ML-DSA support for C# wrapper

tamasan238 opened the pull request:
# Description

Implementing ML-KEM and ML-DSA APIs in the C# wrapper.
For this PR to work, you need to merge other PRs that you will add in comments.

# Testing

Tested.

# Checklist

 - [x] added tests
 - [x] updated/added doxygen
 - [x] updated appropriate READMEs
 - [ ] Updated manual and documentation

tamasan238 commented:
Add _new/_delete API for ML-KEM/ML-DSA
https://github.com/wolfSSL/wolfssl/pull/9039

**Use case**: Add C# examples using ML-KEM/ML-DSA
https://github.com/wolfSSL/wolfssl-examples/pull/514

JacobBarthelmeh reviewed (commented):
@tamasan238 thanks! Looks good for the most part. Please double check the flagged WINCE build comment.

@gojimmypi could you take a look at this PR having C# dev fresh on your mind?

gojimmypi commented:
Hi @tamasan238  - 

Looks like some nice PQ updates for wolfSSL C#. Thank you!

I'm seeing some runtime errors:

```
Starting Cryptographic Tests...

Starting RNG test
RNG Test Passed

Starting ECC tests

[ ... snip other tests ... ]

Comparing Shared Secrets...
Curve25519 shared secret match.

Starting ML-KEM test

Starting ML_KEM_512 shared secret test ...
Testing ML-KEM Key Generation...
Generate Key Pair A...
MlKem key creation exception: System.EntryPointNotFoundException: Unable to find an entry point named 'wc_MlKemKey_New' in DLL 'wolfssl.dll'.
   at wolfSSL.CSharp.wolfcrypt.wc_MlKemKey_New(Int32 type, IntPtr heap, Int32 devId)
   at wolfSSL.CSharp.wolfcrypt.MlKemMakeKey(MlKemTypes type, IntPtr heap, Int32 devId) in C:\PR9040\wolfssl\wrapper\CSharp\wolfSSL_CSharp\wolfCrypt.cs:line 2876
An error occurred: Failed to generate key pair A.

C:\PR9040\wolfssl\wrapper\CSharp\Debug\x64\wolfCrypt-Test.exe (process 26580) exited with code -1 (0xffffffff).
To automatically close the console when debugging stops, enable Tools->Options->Debugging->Automatically close the console when debugging stops.
Press any key to close this window . . .
```


I followed the instructions in the README update:

```
## PQC Suppport

To enable ML-KEM / ML-DSA API, add following options for `wolfssl` and `wolfCrypt-Test` projects:

HAVE_MLKEM
HAVE_MLDSA
```
 <br />

As seen here:

<img width="1087" height="459" alt="image" src="https://github.com/user-attachments/assets/151ce5bb-7fa3-4a56-bb63-b48dfe2b9776" />

<img width="822" height="311" alt="image" src="https://github.com/user-attachments/assets/28b73641-476c-47bd-9696-d6e9ed490096" />
 <br />
 <br />

It appears `wc_MlKemKey_New` is not in the wolfSSL codebase? Perhaps you meant `wc_MlKemKey_Init` or `wc_MlKemKey_MakeKey` ?

<img width="596" height="205" alt="image" src="https://github.com/user-attachments/assets/965451eb-8b54-4a23-ab45-fe2958aa53be" />

<br />

## Unrelated Warnings

There are some warnings that seem to have snuck in over time, unrelated to this PR:

```
1>C:\PR9040\wolfssl\wolfssl\error-ssl.h(245,1): warning C5287: operands are different enum types 'wolfCrypt_ErrorCodes' and 'wolfSSL_ErrorCodes'; use an explicit cast to silence this warning
1>(compiling source file '../../src/crl.c')
1>dtls13.c
1>C:\PR9040\wolfssl\wolfssl\error-ssl.h(245,1): warning C5287: operands are different enum types 'wolfCrypt_ErrorCodes' and 'wolfSSL_ErrorCodes'; use an explicit cast to silence this warning
1>(compiling source file '../../src/dtls13.c')
1>dtls.c
1>C:\PR9040\wolfssl\wolfssl\error-ssl.h(245,1): warning C5287: operands are different enum types 'wolfCrypt_ErrorCodes' and 'wolfSSL_ErrorCodes'; use an explicit cast to silence this warning
1>(compiling source file '../../src/dtls.c')
1>internal.c
1>C:\PR9040\wolfssl\wolfssl\error-ssl.h(245,1): warning C5287: operands are different enum types 'wolfCrypt_ErrorCodes' and 'wolfSSL_ErrorCodes'; use an explicit cast to silence this warning
1>(compiling source file '../../src/internal.c')
1>wolfio.c
1>C:\PR9040\wolfssl\wolfssl\error-ssl.h(245,1): warning C5287: operands are different enum types 'wolfCrypt_ErrorCodes' and 'wolfSSL_ErrorCodes'; use an explicit cast to silence this warning
1>(compiling source file '../../src/wolfio.c')
1>keys.c
1>C:\PR9040\wolfssl\wolfssl\error-ssl.h(245,1): warning C5287: operands are different enum types 'wolfCrypt_ErrorCodes' and 'wolfSSL_ErrorCodes'; use an explicit cast to silence this warning
1>(compiling source file '../../src/keys.c')
1>ocsp.c
1>ssl.c
1>C:\PR9040\wolfssl\wolfssl\error-ssl.h(245,1): warning C5287: operands are different enum types 'wolfCrypt_ErrorCodes' and 'wolfSSL_ErrorCodes'; use an explicit cast to silence this warning
1>(compiling source file '../../src/ssl.c')
1>tls.c
1>C:\PR9040\wolfssl\wolfssl\error-ssl.h(245,1): warning C5287: operands are different enum types 'wolfCrypt_ErrorCodes' and 'wolfSSL_ErrorCodes'; use an explicit cast to silence this warning
1>(compiling source file '../../src/tls.c')
1>tls13.c
1>C:\PR9040\wolfssl\wolfssl\error-ssl.h(245,1): warning C5287: operands are different enum types 'wolfCrypt_ErrorCodes' and 'wolfSSL_ErrorCodes'; use an explicit cast to silence this warning
1>(compiling source file '../../src/tls13.c')
1>aes.c
1>arc4.c
1>asn.c
1>C:\PR9040\wolfssl\wolfssl\error-ssl.h(245,1): warning C5287: operands are different enum types 'wolfCrypt_ErrorCodes' and 'wolfSSL_ErrorCodes'; use an explicit cast to silence this warning
1>(compiling source file '../../wolfcrypt/src/asn.c')
```

I'm addressing those warnings in https://github.com/wolfSSL/wolfssl/pull/8946

tamasan238 commented:
Thank you for your review!
I'll correct each item and contact again.

tamasan238 commented:
1️⃣
Fixed below.

**1. typo**

- s/Suppport/Support/
- s/encalsulate/encapsulate/

**2. WinCE Error**

- type, key, ... -> key, type, ...

**3. configuration options**

added some options for wolfSSL(wolfCrypt).
// I think I wrote it before, but it seems that I blew it away when rebasing git commit. I'm sorry.

2️⃣
These APIs are currently being added in a separate PR. (https://github.com/wolfSSL/wolfssl/pull/9039)

- wolfCrypt/wc_mlkem.c: 
    - wc_MlKemKey_New()
    - wc_MlKemKey_Delete()
- wolfCrypt/wc_dilithium.c: 
    - wc_dilithium_new()
    - wc_dilithium_delete()

I would appreciate it if you could check it after referring to the above.

tamasan238 commented:
Thank you for your review!

Yes, I think it is better to be available without any macros.
But in my understanding, some errors may occur when a user builds wolfSSL without PQC macros.

If some PQC APIs are unavailable in wolfSSL, the C# wrapper test program(`wolfCrypt-Test.cs`) will output an error.
So, I think that at least `wolfCrypt-Test.cs` requires `ifdef` (`if` in C#).
In order to avoid confusing users, I thought it would be better to apply the same style in `wolfCrypt.cs`.

However, I believe it is also important to keep the code base and build procedures as simple as possible.
I would like to hear your perspective.

tamasan238 commented:
Oh, I see! That is the best plan.
I'll do that. Thank you :)

tamasan238 commented:
jenkins retest this please

tamasan238 commented:
Thank you for your confirmation!

> Issue where `wolfSSLCbIoSend()` fails under specific conditions

I still haven't figured out the root cause of this issue.

> Regarding `throw new Exception`

Indeed, even if a test fails, subsequent tests should still be executed.
I also think it would be better to address this in a separate PR, so can we put this PR on hold for now?
To be honest, I'm not really sure why it's implemented this way.

https://github.com/wolfSSL/wolfssl/pull/9039 should be merged soon.
After confirming this, I will rerun the GitHub Actions/Jenkins tests and get back to you.

tamasan238 commented:
The API addition PR(https://github.com/wolfSSL/wolfssl/pull/9039) has been merged!
And by rebasing, I confirmed the following:
- The haproxy test issue now passes. (https://github.com/wolfSSL/wolfssl/pull/9061)
- Using the added API, `wolfCrypt-Test.cs` now succeeds.

Could you please re-check or/and merge this PR?

gojimmypi commented:
I've confirmed the ML-KEM / ML-DSA test app runs successfully in VS 2022. 

I didn't get a chance today to test client / server apps, nor a specific code review yet.

```
[ ... snip ... ]
Starting ML_KEM_1024 shared secret test ...
Testing ML-KEM Key Generation...
Generate Key Pair A...
Generate Key Pair B...
ML-KEM Key Generation test passed.
Testing ML-KEM Key Encode...
ML-KEM Key Encode test passed.
Testing ML-KEM Encapsulation...
ML-KEM Encapsulation test passed.
Testing ML-KEM Decode...
ML-KEM Decode test passed.
Testing ML-KEM Decapsulation...
ML-KEM Decapsulation test passed.
Comparing Shared Secrets...
ML-KEM shared secret match.

Starting ML-DSA test

Starting ML_DSA_44 shared secret test ...
Testing ML-DSA Key Generation...
ML-DSA Key Generation test passed.
Testing ML-DSA Key Export...
ML-DSA Key Export test passed.
Testing ML-DSA Key Import...
ML-DSA Key Import test passed.
Testing ML-DSA Signature Creation...
ML-DSA Signature Creation test passed. Signature Length: 2420
Testing ML-DSA Signature Verification...
ML-DSA Signature Verification test passed.
[ ... snip ... ]
```

tamasan238 commented:
jenkins retest this please

tamasan238 commented:
Thank you for checking.
There are still many things that need to be fixed. I'm learning a lot from you explaining your thinking in detail.
I'll continue working on the fixes.

tamasan238 commented:
I've made several fixes and responded to some review comments.
Please review the changes.

Note that the system drive on my development Windows machine has failed, so I'm unable to run tests locally using Visual Studio.
As a result, there may be some basic errors present.
I plan to fix these by referencing the test results from Jenkins/GitHub Actions.
If you notice any issues, I'd appreciate it if you could let me know.

tamasan238 commented:
Additionally, I was advised via Slack to refer to the following regarding error values, but it appears they are not referenced in `wolfCrypt.cs`.

https://github.com/wolfSSL/wolfssl/blob/master/wolfssl/wolfcrypt/error-crypt.h

Please let me know what error values you consider desirable to set, or whether you think a reference to `error-crypt.h` should be added.

```
"D:\a\wolfssl\wolfssl\wolfssl\wrapper\CSharp\wolfSSL_CSharp.sln" (default target) (1) ->
"D:\a\wolfssl\wolfssl\wolfssl\wrapper\CSharp\wolfSSL-TLS-PSK-Server\wolfSSL-TLS-PSK-Server.csproj" (default target) (6) ->
"D:\a\wolfssl\wolfssl\wolfssl\wrapper\CSharp\wolfSSL_CSharp\wolfSSL_CSharp.csproj" (default target) (10:9) ->
(CoreCompile target) -> 
 D:\a\wolfssl\wolfssl\wolfssl\wrapper\CSharp\wolfSSL_CSharp\wolfCrypt.cs(2960,24): error CS0103: The name 'BAD_FUNC_ARG' does not exist in the current context [D:\a\wolfssl\wolfssl\wolfssl\wrapper\CSharp\wolfSSL_CSharp\wolfSSL_CSharp.csproj]
 D:\a\wolfssl\wolfssl\wolfssl\wrapper\CSharp\wolfSSL_CSharp\wolfCrypt.cs(3004,24): error CS0103: The name 'BAD_FUNC_ARG' does not exist in the current context [D:\a\wolfssl\wolfssl\wolfssl\wrapper\CSharp\wolfSSL_CSharp\wolfSSL_CSharp.csproj]
 D:\a\wolfssl\wolfssl\wolfssl\wrapper\CSharp\wolfSSL_CSharp\wolfCrypt.cs(3047,24): error CS0103: The name 'BAD_FUNC_ARG' does not exist in the current context [D:\a\wolfssl\wolfssl\wolfssl\wrapper\CSharp\wolfSSL_CSharp\wolfSSL_CSharp.csproj]
 D:\a\wolfssl\wolfssl\wolfssl\wrapper\CSharp\wolfSSL_CSharp\wolfCrypt.cs(3087,24): error CS0103: The name 'BAD_FUNC_ARG' does not exist in the current context [D:\a\wolfssl\wolfssl\wolfssl\wrapper\CSharp\wolfSSL_CSharp\wolfSSL_CSharp.csproj]
 D:\a\wolfssl\wolfssl\wolfssl\wrapper\CSharp\wolfSSL_CSharp\wolfCrypt.cs(3132,24): error CS0103: The name 'BAD_FUNC_ARG' does not exist in the current context [D:\a\wolfssl\wolfssl\wolfssl\wrapper\CSharp\wolfSSL_CSharp\wolfSSL_CSharp.csproj]
 D:\a\wolfssl\wolfssl\wolfssl\wrapper\CSharp\wolfSSL_CSharp\wolfCrypt.cs(3184,24): error CS0103: The name 'BAD_FUNC_ARG' does not exist in the current context [D:\a\wolfssl\wolfssl\wolfssl\wrapper\CSharp\wolfSSL_CSharp\wolfSSL_CSharp.csproj]
 D:\a\wolfssl\wolfssl\wolfssl\wrapper\CSharp\wolfSSL_CSharp\wolfCrypt.cs(3327,24): error CS0103: The name 'BAD_FUNC_ARG' does not exist in the current context [D:\a\wolfssl\wolfssl\wolfssl\wrapper\CSharp\wolfSSL_CSharp\wolfSSL_CSharp.csproj]
 D:\a\wolfssl\wolfssl\wolfssl\wrapper\CSharp\wolfSSL_CSharp\wolfCrypt.cs(3351,24): error CS0103: The name 'BAD_FUNC_ARG' does not exist in the current context [D:\a\wolfssl\wolfssl\wolfssl\wrapper\CSharp\wolfSSL_CSharp\wolfSSL_CSharp.csproj]
 D:\a\wolfssl\wolfssl\wolfssl\wrapper\CSharp\wolfSSL_CSharp\wolfCrypt.cs(3380,24): error CS0103: The name 'BAD_FUNC_ARG' does not exist in the current context [D:\a\wolfssl\wolfssl\wolfssl\wrapper\CSharp\wolfSSL_CSharp\wolfSSL_CSharp.csproj]
 D:\a\wolfssl\wolfssl\wolfssl\wrapper\CSharp\wolfSSL_CSharp\wolfCrypt.cs(3430,24): error CS0103: The name 'BAD_FUNC_ARG' does not exist in the current context [D:\a\wolfssl\wolfssl\wolfssl\wrapper\CSharp\wolfSSL_CSharp\wolfSSL_CSharp.csproj]
 D:\a\wolfssl\wolfssl\wolfssl\wrapper\CSharp\wolfSSL_CSharp\wolfCrypt.cs(3482,24): error CS0103: The name 'BAD_FUNC_ARG' does not exist in the current context [D:\a\wolfssl\wolfssl\wolfssl\wrapper\CSharp\wolfSSL_CSharp\wolfSSL_CSharp.csproj]
 D:\a\wolfssl\wolfssl\wolfssl\wrapper\CSharp\wolfSSL_CSharp\wolfCrypt.cs(3531,24): error CS0103: The name 'BAD_FUNC_ARG' does not exist in the current context [D:\a\wolfssl\wolfssl\wolfssl\wrapper\CSharp\wolfSSL_CSharp\wolfSSL_CSharp.csproj]
 ```

tamasan238 commented:
Additionally, I note that the current code includes code to force the test to terminate prematurely, as suggested to me via Slack.

```
/* DEBUG CODE FOR REVIEW */
ret = BAD_FUNC_ARG;
```

tamasan238 commented on wrapper/CSharp/wolfCrypt-Test/wolfCrypt-Test.cs:
```diff
@@ -648,6 +648,201 @@ private static void curve25519_test()
+                Console.WriteLine("ML-KEM shared secret match.");
+            }
+        }
+        finally
```
It's related to https://github.com/wolfSSL/wolfssl/pull/9040#discussion_r2462588258

gojimmypi commented:
> Additionally, I note that the current code includes code to force the test to terminate prematurely, as suggested to me via Slack.
> 
> ```
> /* DEBUG CODE FOR REVIEW */
> ret = BAD_FUNC_ARG;
> ```

See the error codes for `BAD_FUNC_ARG` and others in [wolfssl/wolfcrypt/error-crypt.h](https://github.com/wolfSSL/wolfssl/blob/master/wolfssl/wolfcrypt/error-crypt.h); it might be nice to have them all available in the C# wrapper, or you could just define the ones you use. Try to use consistent values if not importing the entire list.

gojimmypi commented:
> Note that the system drive on my development Windows machine has failed, so I'm unable to run tests locally using Visual Studio. As a result, there may be some basic errors present.

Any luck on your Visual Studio? I'm seeing these compile-time errors:

```
Use of unassigned local variable 'privA'
Use of unassigned local variable 'pubA'
Use of unassigned local variable 'cipherText'
The name 'BAD_FUNC_ARG' does not exist in the current context
Use of unassigned local variable 'sharedSecretA'
Use of unassigned local variable 'sharedSecretB'
Use of unassigned local variable 'privateKey'
Use of unassigned local variable 'publicKey'
Use of unassigned local variable 'signature'
```

gojimmypi commented on wrapper/CSharp/wolfCrypt-Test/wolfCrypt-Test.cs:
```diff
@@ -648,6 +648,289 @@ private static void curve25519_test()
+            }
+            if (keyB != IntPtr.Zero)
+            {
+                ret = wolfcrypt.MlKemFreeKey(keyB);
```
Any prior `ret` error values are lost and overwritten here. Is this as intended?

gojimmypi commented on wrapper/CSharp/wolfSSL_CSharp/wolfCrypt.cs:
```diff
@@ -2709,6 +2850,658 @@ public static void Curve25519ExportKeyRaw(IntPtr key, out byte[] privateKey, out
+                ret = wc_dilithium_delete(key, ref heap);
+                key = IntPtr.Zero;
+            }
+            return ret;
```
yes, see my design-pattern comment there; point being is to make it easy for the reader to know the default return code is intentional.

tamasan238 commented:
Once preparations are complete, I will create a new PR.
